### PR TITLE
Dht22 measurement

### PR DESCRIPTION
**Pull request by miczyg1**

### CODE REVIEW

**Comment by macpijan** [comment]
@miczyg1 Take a look at this PR: it includes many more commits than your post. Blog posts PR should go into `source` branch. This should fix the issue

**Comment by pietrushnic** [comment]
@miczyg1 please rebase those changes and create PR with merge request to `source` branch.

File: source/_posts/2017-08-02-dht22.markdown [hunk 1]
```diff
@@ -0,0 +1,214 @@
+---
+author: Michał Żygowski
+layout: post
+title: "1-Wire read data protocol on DHT22 sensor"
+date: 2017-08-1 14:14:55 +0200
+comments: true
+categories: sensors measurements DHT22 1-Wire mbed
+---
+
+Recently I have encountered with temperature and humidity measurements using DHT22 sensor. I was developing a driver source code in ARM mbed OS SDK on particular STM32 NUCLEO L432KC platform. Thorough analysis of DHT22 [documentation](https://cdn-shop.adafruit.com/datasheets/Digital+humidity+and+temperature+sensor+AM2302.pdf) led me to the following questions:
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 please wrap at 80 characters

File: source/_posts/2017-08-02-dht22.markdown [hunk 2]
```diff
@@ -0,0 +1,214 @@
+---
+author: Michał Żygowski
+layout: post
+title: "1-Wire read data protocol on DHT22 sensor"
+date: 2017-08-1 14:14:55 +0200
+comments: true
+categories: sensors measurements DHT22 1-Wire mbed
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 please add stm32, all tags should be lower case

File: source/_posts/2017-08-02-dht22.markdown [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+---
+
+Recently I have encountered with temperature and humidity measurements using DHT22 sensor. I was developing a driver source code in ARM mbed OS SDK on particular STM32 NUCLEO L432KC platform. Thorough analysis of DHT22 [documentation](https://cdn-shop.adafruit.com/datasheets/Digital+humidity+and+temperature+sensor+AM2302.pdf) led me to the following questions:
+
+- Is it possible to accurately measure voltage-level durations during read process?
+- What duration time values should be considered as timeout or/and error?
+- Should I weaken the time restrictions in order to avoid random delays in voltage level transitions be considered as failure?
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 please check whole file against 80 character wrap.

File: source/_posts/2017-08-02-dht22.markdown [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+For ARM mbed OS documentation please refer to [mbed API documentation](https://developer.mbed.org/users/screamer/code/mbed/docs/tip/)
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 please do not wrap links this breaks markdown parser.

File: source/_posts/2017-08-02-dht22.markdown [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+### Configuration
+
+Let's start with a little bit of configuration and statistics.
+
+The STM32 NUCLEO L432KC is cocked with up to 80MHz frequency which gives a period of 125 nanoseconds, so basically 8 periods sums to 1 microsecond.
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 `cocked` ? It has connotations

File: source/_posts/2017-08-02-dht22.markdown [hunk 6]
```diff
@@ -0,0 +1,214 @@
+---
+author: Michał Żygowski
+layout: post
+title: "1-Wire read data protocol on DHT22 sensor"
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 maybe something more human and SEO freindly: How to handle DHT22 sensor using ARM mbed OS ?

There may be something different but please consider what other people will write into search engine to find your post.

File: source/_posts/2017-08-02-dht22.markdown [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+tmr.reset();
+if(n != 80) {
+    tmr.stop();
+    return DHT_RESPONSE_ERROR;
+}
+```
+At this point we can deliberate about the `TIMEOUT` value and the time restrictions provided in the `if` expressions. 100 microseconds seemst t be reasonable value for `TIMEOUT`, beacause there is no voltage level duration longer than 80 microsecond defined by the protocol. 
+
+Running this code will certainly lead to returning `DHT_RESPONSE_ERROR`. Why? The timer restriction provided in `if` expressions are too strict. Tests conducted by me showed, that `timer` does not always read the same amount of microseconds passed each time i ran this code. The values fluctuated between 70 and even 90 microseconds. This dispersion is unacceptable considering 125 nanoseconds clock period in STM32 NUCLEO L432KC. It inspired me to investigate the hardware layer for possible faults. I have used the Saleae Logic Analyzer to monitor the sensor's data line. The result occured to be litttle surprising. The data line waveform I captured is showed below. The sampling frequency was set to 12MHz.
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 how I can comment something here without wrap ?

File: source/_posts/2017-08-02-dht22.markdown [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    }
+} while(dht_data.read() == 1);
+
+tmr.reset();
+if(n != 80) {
+    tmr.stop();
+    return DHT_RESPONSE_ERROR;
+}
+```
+At this point we can deliberate about the `TIMEOUT` value and the time restrictions provided in the `if` expressions. 100 microseconds seemst t be reasonable value for `TIMEOUT`, beacause there is no voltage level duration longer than 80 microsecond defined by the protocol.
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 `seems to be`

**Review by pietrushnic — CHANGES_REQUESTED** [review]

File: source/_posts/2017-08-02-dht22.markdown [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+```
+At this point we can deliberate about the `TIMEOUT` value and the time
+restrictions provided in the `if` expressions. 100 microseconds seems to be
+reasonable value for `TIMEOUT`, beacause there is no voltage level duration 
+longer than 80 microsecond defined by the protocol. 
+
+Running this code will certainly lead to returning `DHT_RESPONSE_ERROR`. Why?
+The timer restriction provided in `if` expressions are too strict. Tests
+conducted by me showed, that `timer` does not always read the same amount of
+microseconds passed each time i ran this code. The values fluctuated between 70
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 `I ran`

File: source/_posts/2017-08-02-dht22.markdown [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+and even 90 microseconds. This dispersion is unacceptable considering 125
+nanoseconds clock period in STM32 NUCLEO L432KC. It inspired me to investigate
+the hardware layer for possible faults. I have used the Saleae Logic Analyzer to
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 let's avoid vendor name, jus logic analyzer

File: source/_posts/2017-08-02-dht22.markdown [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+monitor the sensor's data line. The result occured to be litttle surprising. The
+data line waveform I captured is showed below. The sampling frequency was set to
+12MHz.
+
+<a class="fancybox" rel="group" href="/assets/images/dht22-response.png"><img src="/assets/images/dht22-response.png" alt="" /></a>
+
+The sensor pulled the line down after c.a. 22 microseconds, which is apropiate.
+But then the voltage level durations differ slighlty, they are 1.5 microseconds
+far from 80. In few cases I observed also a 90 microseconds long low voltage
+level.To confirm the realiability of this measurement I have additionally
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 `level. To`

File: source/_posts/2017-08-02-dht22.markdown [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+connected the data line to RIGOL DS1104 oscilloscope. The results were the same
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 just oscilloscope

File: source/_posts/2017-08-02-dht22.markdown [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+            return DHT_READ_BIT_ERROR;
+        }
+
+        do {
+            n = timer.read_us();
+            if(n > TIMEOUT) {
+                timer.stop();
+                return DHT_READ_BIT_TIMEOUT;
+            }
+        } while(dht_data.read() == 1);
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown:201]: @miczyg1 repeated code can be moved to separate function

File: source/_posts/2017-08-02-dht22.markdown [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+sensor sooner than 2 seconds since last poll.
+
+
+## Summary
+
+Above steps gave me ability to take temperature and humidity measurements by
+DHT22. This particular sensor has its pros as well as cons.
+
+Its advantage certainly is simplicity in hardware design. Only a pull up
+resistor, filtering condensator and 3 wire connections are needed.
```
Comment (pietrushnic) [inline source/_posts/2017-08-02-dht22.markdown]: @miczyg1 price ? popularity ?

**Review by pietrushnic — CHANGES_REQUESTED** [review]

File: source/_posts/2017-08-02-dht22.markdown (hunk 13 above)
Comment (miczyg1) [reply source/_posts/2017-08-02-dht22.markdown:201]: These do..while loops differ from each other, they return  read stage dependent error numbers.

**Pull request merged**